Rewrite particle output to use deal.II functionality

gassmoeller opened the pull request:
This PR uses the recently introduced funtionality of deal.II (dealii/dealii#308) to output zero-dimensional output features (like particles). It essentially offers the same functionality that `Visualization` already implements, but for the particles. With this PR the `particle/output/` plugins will no longer be necessary once we require deal.II 9.0. @hlokavarapu: This will conflict with #1049. Are you still interested in that functionality? Then we should think about integrating that behavior into this PR.

Things still to do:
- [x] make sure existing tests are not broken with deal.II 8.5
- [x] write a changelog entry

bangerth commented on include/aspect/postprocess/particles.h:
```diff
@@ -27,10 +27,79 @@
+       * of dim (the dimension in which this zero-dimensional particle lives).
+       */
+      template<int dim>
+      class ParticleOutput : public dealii::DataOutInterface<0,dim>
```
Don't you have to guard this class with `#if DEAL_II_VERSION_GTE(9,0)` as well?

bangerth commented on source/postprocess/particles.cc:
```diff
@@ -31,21 +32,119 @@ namespace aspect
+    {
+      template<int dim>
+      ParticleOutput<dim>::~ParticleOutput()
+      {}
```
do you need this destructor?

bangerth commented on source/postprocess/particles.cc:
```diff
@@ -31,21 +32,119 @@ namespace aspect
+                dataset_names.push_back(field_name);
+            else
+              for (unsigned int component_index=0; component_index<n_components; ++component_index)
+                dataset_names.push_back(field_name + "[" + Utilities::to_string(component_index) + "]");
```
I don't think `[` and `]` are valid names for data set names. You may have to just append the index with an underscore.

bangerth commented on source/postprocess/particles.cc:
```diff
@@ -31,21 +32,119 @@ namespace aspect
+            for (unsigned int property_index = 0; property_index < properties.size(); ++property_index)
+              patches[i].data(property_index+1,0) = properties[property_index];
+
+          }
```
no need for the empty line

gassmoeller commented:
Yes, I did not check the old functionality. This is fixed now (should work with 8.5.0 and 9.0.pre). I also removed the empty destructor and updated the test results.

hlokavarapu commented:
Woah, I have been waiting for this day; compressed particle visualization data, oh man pretty excited :). I don;t want to slow things down but I am still interested in supporting writing out multiple formats for particles. I will export the relevant code of PR #1049 to this code branch. This will take me a day or two.

gassmoeller commented:
Sounds good. From a first glance I would think it should fit neatly into the new structure, if you make output_format a list/vector of unique formats, and then loop over the if/else if branches in particles.cc:357-492. I will check in the meantime what happens with the test failures.

gassmoeller commented:
This should be good to go. Supersedes #1049.

bangerth commented on source/postprocess/particles.cc:
```diff
@@ -31,21 +32,119 @@ namespace aspect
+            patches[i].n_subdivisions = 1;
+            patches[i].data.reinit(property_information.n_components()+1,1);
+
+            patches[i].data(0,0) = particle->second.get_id();
```
We must be building the worst-suited data structure to represent particle data here :-) But that's what it is, nothing that can be done about it ;-)

bangerth replied:
I think it might be worthwhile specializing `Patch<0,spacedim>` to avoid some of the member variables. But that can wait.

bangerth replied:
See https://github.com/dealii/dealii/issues/4695

gassmoeller commented:
I addressed the comments. Should be ready to go.
